File: week08/15/4_Thu_MinStack.cpp
#include <stack>

using namespace std;

class MinStack {
public:
    stack<int> mainstack;
    stack<int> minstack;

    MinStack() {
        
    }
    
    void push(int val) {
        mainstack.push(val);
        if (minstack.empty() || val <= minstack.top())
            minstack.push(val); 
    }
    
    void pop() {
        if (!mainstack.empty()) { 
            if (!minstack.empty() && mainstack.top() == minstack.top()) minstack.pop();
            mainstack.pop();
        }
    }
    
    int top() {
        return mainstack.top();
    }
    
    int getMin() {
        return minstack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: week08/15/4_Thu_MinStack.py
class MinStack:

    def __init__(self):
        self.mainstack = []
        self.minstack = []

    def push(self, val: int) -> None:
        self.mainstack.append(val)
        
        if not self.minstack or val <= self.minstack[-1]:
            self.minstack.append(val)

    def pop(self) -> None:
        if self.mainstack:
            if self.mainstack[-1] == self.minstack[-1]:
                self.minstack.pop()
            self.mainstack.pop()        

    def top(self) -> int:
        if self.mainstack:
            return self.mainstack[-1]

    def getMin(self) -> int:
        if self.minstack:
            return self.minstack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()